File: elixir/binary-search/lib/binary_search.ex
defmodule BinarySearch do
  @doc """
    Searches for a key in the tuple using the binary search algorithm.
    It returns :not_found if the key is not in the tuple.
    Otherwise returns {:ok, index}.

    ## Examples

      iex> BinarySearch.search({}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 2)
      :not_found

      iex> BinarySearch.search({1, 3, 5}, 5)
      {:ok, 2}

  """

  @spec search(tuple, integer) :: {:ok, integer} | :not_found
  def search(numbers, key) do
    do_search(numbers, key, 0, tuple_size(numbers) - 1)
  end

  @spec do_search(tuple, integer, non_neg_integer, non_neg_integer) :: {:ok, integer} | :not_found
  defp do_search(_, _, bot, top) when bot > top, do: :not_found
  defp do_search(_, _, bot, top) when top < bot, do: :not_found

  defp do_search(numbers, key, bot, top) do
    middle = floor((bot + top) / 2)
    current = elem(numbers, middle)

    cond do
      current == key -> {:ok, middle}
      current < key -> do_search(numbers, key, bot + 1, top)
      true -> do_search(numbers, key, bot, top - 1)
    end
  end
end

File: php/binary-search/BinarySearch.php
<?php

declare(strict_types=1);

/**
 * Performs a binary search to find the index of a target value within a sorted array.
 *
 * This function efficiently searches for a `$needle` within a `$haystack` array.
 * It repeatedly divides the search interval in half. If the value is found, its index is returned.
 * If the value is not present in the array, -1 is returned.
 * The input array MUST be sorted for this algorithm to work correctly.
 *
 * @param int $needle The value to search for.
 * @param array<int> $haystack The sorted array to search within.
 * @return int The index of the `$needle` if found, otherwise -1.
 *
 * @example
 * find(6, [6]);                     // Returns 0
 * find(6, [1, 3, 4, 6, 8, 9, 11]);  // Returns 3
 * find(1, [1, 3, 4, 6, 8, 9, 11]);  // Returns 0
 * find(11, [1, 3, 4, 6, 8, 9, 11]); // Returns 6
 * find(7, [1, 3, 4, 6, 8, 9, 11]);  // Returns -1
 * find(1, []);                      // Returns -1
 */
function find(int $needle, array $haystack): int
{
    $bot = 0;
    $top = count($haystack) - 1;

    for ($middle = middleFor($bot, $top); $bot <= $top; $middle = middleFor($bot, $top)) {
        $element = $haystack[$middle];

        if ($needle === $element) {
            return $middle;

        } else if ($needle < $element) {
            $top = $middle - 1;

        } else {
            $bot = $middle + 1;
        }
    }

    return -1;
}

/**
 * Calculates the middle index between a 'bottom' and 'top' boundary.
 *
 * This is a helper function for the binary search algorithm, ensuring the middle index
 * is always an integer by rounding down.
 *
 * @param int $bot The lower bound of the interval.
 * @param int $top The upper bound of the interval.
 * @return int The integer middle index.
 */
function middleFor(int $bot, int $top): int
{
    $avg = ($bot + $top) / 2;
    return (int) (floor($avg));
}